rollup/ccc/async_checker.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/scroll-tech/go-ethereum/core/state"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/core/vm"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/metrics"
+	"github.com/scroll-tech/go-ethereum/params"
+	"github.com/scroll-tech/go-ethereum/rollup/tracing"
+)
+
+var (
+	failCounter        = metrics.NewRegisteredCounter("ccc/async/fail", nil)
+	checkTimer         = metrics.NewRegisteredTimer("ccc/async/check", nil)
+	activeWorkersGauge = metrics.NewRegisteredGauge("ccc/async/active_workers", nil)
+)
+
+type Blockchain interface {
+	Database() ethdb.Database
+	GetBlock(hash common.Hash, number uint64) *types.Block
+	StateAt(root common.Hash) (*state.StateDB, error)
+	Config() *params.ChainConfig
+	GetVMConfig() *vm.Config
+	CurrentHeader() *types.Header
+	core.ChainContext
+}
+
+// AsyncChecker allows a caller to spawn CCC verification tasks
+type AsyncChecker struct {
+	bc             Blockchain
+	onFailingBlock func(*types.Block, error)
+
+	workers      *stream.Stream
+	freeCheckers chan *Checker
+
+	// local state to keep track of the chain progressing and terminate tasks early if needed
+	currentHead       *types.Header
+	forkCtx           context.Context
+	forkCtxCancelFunc context.CancelFunc
+}
+
+type ErrorWithTxnIdx struct {
+	txIdx      uint
+	err        error
+	shouldSkip bool
+}
+
+func (e *ErrorWithTxnIdx) Error() string {
+	return fmt.Sprintf("txn at index %d failed with %s", e.txIdx, e.err)
+}
+
+func (e *ErrorWithTxnIdx) Unwrap() error {
+	return e.err
+}
+
+func NewAsyncChecker(bc Blockchain, numWorkers int, lightMode bool) *AsyncChecker {
+	forkCtx, forkCtxCancelFunc := context.WithCancel(context.Background())
+	return &AsyncChecker{
+		bc: bc,
+		freeCheckers: func(count int) chan *Checker {
+			checkers := make(chan *Checker, count)
+			for i := 0; i < count; i++ {
+				checkers <- NewChecker(lightMode)
+			}
+			return checkers
+		}(numWorkers),
+		workers:           stream.New().WithMaxGoroutines(numWorkers),
+		currentHead:       bc.CurrentHeader(),
+		forkCtx:           forkCtx,
+		forkCtxCancelFunc: forkCtxCancelFunc,
+	}
+}
+
+func (c *AsyncChecker) WithOnFailingBlock(onFailingBlock func(*types.Block, error)) *AsyncChecker {
+	c.onFailingBlock = onFailingBlock
+	return c
+}
+
+func (c *AsyncChecker) Wait() {
+	c.workers.Wait()
+}
+
+// Check spawns an async CCC verification task.
+func (c *AsyncChecker) Check(block *types.Block) error {
+	if block.NumberU64() > c.currentHead.Number.Uint64()+1 {
+		log.Error("non continuous chain observed in AsyncChecker", "prev", c.currentHead, "got", block.Header())
+	} else if block.ParentHash() != c.currentHead.Hash() {
+		// seems like there is a fork happening, a block from the canonical chain must have failed CCC check
+		// assume the incoming block is the new tip in the fork
+		c.forkCtx, c.forkCtxCancelFunc = context.WithCancel(context.Background())
```

Comment:
Does this line implicitly cancel the previous fork context?

Reply:
Nope, it doesnt. But if we are observing a fork here, it is most likely that previous fork context was canceled already.